client/tinychain/ml/linalg.py

```diff
@@ -273,3 +277,380 @@ def svd(matrix: Tensor) -> Tuple:
+                ))
+
+            @closure(txn.n)
+            @post_op
+            def cond(j: UInt) -> Bool:
+                return j < txn.n
+
+            cxt.update_g = 1.0 / g
+            loop = After(cxt.update_g, While(cond, step, Map(j=l, n=txn.n, l=l, i=i)))
+            update_U = After([loop, U[i:,i].write(U[i:,i]*cxt.update_g)], cxt.update_g)
+
+            return If((g != 0.0), update_U, g)
+
+        cxt.update_cols = update_cols
+        cxt.new_l = UInt(i + 1)
+        cxt.l = UInt(After(U[i, cxt.new_l:].write(0.0), cxt.new_l))
+        cxt.new_g = Float(W[i])
+        cxt.final_g = cxt.update_cols(i=i, l=cxt.l, g=cxt.new_g)
+
+        return Map(i=i - 1, l=cxt.l, g=cxt.final_g, U=U, W=W)
+
+    @post_op
+    def cond(i: UInt) -> Bool:
+        return i >= 0
+
+    return Map(While(cond, step, Map(i=txn.min_m_n-1, l=0, g=0.0, U=U, W=W)))
+
+
+# helper functions
+def pythag(a,b): 
+    aa = a.abs()
+    bb = b.abs()
+    return If(aa > bb, aa * (1.0 + (bb/aa)**2)**0.5, bb * (1.0 + (aa/bb)**2)**0.5)
+
+
+@post_op
+def cancelation(txn, U, W, e, l, k):
+    txn.c = 0.0
+    txn.s = 1.0
+    txn.l1 = l-1
+
+    @closure(U, W, e, l, k, txn.c, txn.s, txn.l1)
+    @post_op
+    def step(cxt, i: UInt) -> Map:
+        cxt.f = txn.s * e[i]
+        e[i] = txn.c * e[i]
+        cxt.g = W[i]
+        cxt.h = pythag(cxt.f, cxt.g)
+        W[i] = cxt.h
+        txn.c = cxt.g/cxt.h
+        txn.s  = -cxt.f/cxt.h
+        cxt.Y = U[:,txn.l1].copy()
+        cxt.Z = U[:,i].copy()
+        update_prev_col = U[:,txn.l1].write(cxt.Y * txn.c + cxt.Z * txn.s)
+        update_col = U[:,i].write(-cxt.Y * txn.s + cxt.Z * txn.c)
+        U_final = After([update_prev_col, update_col], U)
+        return If(cxt.f.abs() <= EPS, 
+                Map(running=cxt.f.abs() <= EPS, i=i + 1, U=U_final, W=W, e=e),
+                Map(running=cxt.f.abs() <= EPS, i=i + 1, U=U, W=W, e=e))
+
+    @closure(k)
+    @post_op
+    def cond(running: Bool, i: UInt) -> Bool:
+        return running.logical_and(i < k + 1)
+
+    return Map(While(cond, step, Map(running = True, i=l)))    
+
+
+@post_op
+def testfsplit(txn, U, W, e, k):
+    @closure(U, W, e, k)
+    @post_op
+    def step(cxt, l: UInt) -> Map:
+        goto_test_f_convergence_not = If((e[l].abs() <= EPS), False, True)
+        running = If((W[l-1].abs() <= EPS), False, True)
+        return Map(l=l - 1, goto_test_f_convergence_not=goto_test_f_convergence_not, running=running, U=U, W=W, e=e)
+
+    @closure(k)
+    @post_op
+    def cond(l: UInt, goto_test_f_convergence_not:Bool, running: Bool) -> Bool:
+        return (running.logical_and(goto_test_f_convergence_not)).logical_and(l >= 0)
+
+    txn.res = Map(While(cond, step, Map(l=k, goto_test_f_convergence_not = False)))
+    txn.l = txn.res['l']
+    txn.goto_test_f_convergence_not = txn.res['goto_test_f_convergence_not']
+    txn.cancelation = cancelation(U, W, e, txn.l, k)
+
+    return If(txn.goto_test_f_convergence_not, txn.cancelation, txn.l)
```

Comment:
and this can be `If(txn.res['goto_test_f_convergence'], txn.l, txn.cancelation)`